test/spec/Feature/Query/SpreadQueriesSpec.hs

```diff
@@ -112,3 +91,374 @@ spec =
+          , matchHeaders = [matchContentTypeJson]
+          }
+      it "should work when selecting all columns, aggregating each one of them" $
+        get "/factories?select=factory:name,...processes(*)&id=lte.2&order=name" `shouldRespondWith`
+          [json|[
+            {"factory":"Factory A","id":[1,2],"name":["Process A1","Process A2"],"factory_id":[1,1],"category_id":[1,2]},
+            {"factory":"Factory B","id":[3,4],"name":["Process B1","Process B2"],"factory_id":[2,2],"category_id":[1,1]}
+          ]|]
+          { matchStatus  = 200
+          , matchHeaders = [matchContentTypeJson]
+          }
+      it "should aggregate spread columns from a nested one-to-one relationship" $
+        get "/factories?select=factory:name,...processes(process:name,...process_costs(process_costs:cost))&id=lte.2" `shouldRespondWith`
+          [json|[
+            {"factory":"Factory B","process":["Process B1", "Process B2"],"process_costs":[180.00, 70.00]},
+            {"factory":"Factory A","process":["Process A1", "Process A2"],"process_costs":[150.00, 200.00]}
+          ]|]
+          { matchStatus  = 200
+          , matchHeaders = [matchContentTypeJson]
+          }
+      it "should aggregate spread columns from a nested many-to-one relationship" $
+        get "/factories?select=factory:name,...processes(process:name,...process_categories(categories:name))&id=lte.2" `shouldRespondWith`
+          [json|[
+            {"factory":"Factory B","process":["Process B1", "Process B2"],"categories":["Batch", "Batch"]},
+            {"factory":"Factory A","process":["Process A1", "Process A2"],"categories":["Batch", "Mass"]}
+          ]|]
+          { matchStatus  = 200
+          , matchHeaders = [matchContentTypeJson]
+          }
+      it "should aggregate spread columns from a nested one-to-many relationship" $
+        get "/factories?select=factory:name,...processes(process:name,...process_supervisor(supervisor_ids:supervisor_id))&id=lte.2" `shouldRespondWith`
+          [json|[
+            {"factory":"Factory B","process":["Process B1", "Process B1", "Process B2", "Process B2"],"supervisor_ids":[3, 4, 1, 2]},
+            {"factory":"Factory A","process":["Process A1", "Process A2"],"supervisor_ids":[1, 2]}
+          ]|]
+          { matchStatus  = 200
+          , matchHeaders = [matchContentTypeJson]
+          }
+      it "should aggregate spread columns from a nested many-to-many relationship" $ do
+        get "/factories?select=factory:name,...processes(process:name,...supervisors(supervisors:name))&id=lte.2" `shouldRespondWith`
+          [json|[
+            {"factory":"Factory B","process":["Process B1", "Process B1", "Process B2", "Process B2"],"supervisors":["Peter", "Sarah", "Mary", "John"]},
+            {"factory":"Factory A","process":["Process A1", "Process A2"],"supervisors":["Mary", "John"]}
+          ]|]
+          { matchStatus  = 200
+          , matchHeaders = [matchContentTypeJson]
+          }
```

Comment:
> The alternative would be to spread this into an array of arrays, right?

Yes, that was another option. In that case, the response for the first test would be something like:

```json
[
  {"factory":"Factory B","process":["Process B1", "Process B2"],"supervisor_ids":[[3, 4], [1, 2]]},
  {"factory":"Factory A","process":["Process A1", "Process A2"],"supervisor_ids":[[1], [2]]}
]
```
A caveat here is that it's difficult to hoist aggregates to the top resource. For example if we wanted a `COUNT` on supervisors, we could expect this:

```
{"factory":"Factory B","process":["Process B1", "Process B2"],"supervisor_ids":[[3, 4], [1, 2]],"supervisor_count":[2,2]}
```

If we wanted to aggregate to the top level we would need to do a `COUNT(supervisors)` and then a `SUM(supervisors_result)` to get the correct `"count": 4` (unless there's another alternative).

Also, [in the issue description ](https://github.com/PostgREST/postgrest/issues/3041#issue-1980382952) under "Multiple Levels" an example of an expected result is given with flatten arrays (although with a junction table).

> This flattening of nested arrays somehow breaks my mental model about this.

Yes, it actually goes away from the concept of `to-one` spread in that sense (removing the object to select the columns). To avoid both of the issues above, the mental model I went for is this one:

- Treat every "to-many" spread as an `array_agg` aggregate of every selected column in the embedded resources. This aggregate will have a single `GROUP BY` at the top resource. 
- Since a single `GROUP BY` is used, other aggregates can be easily hoisted to the top resource and grouped by the columns selected there.

Reply:
Yes, it's less surprising, although a bit confusing on what to expect when there are more than one column, aggregates or nested relationships.

This case is straight forward enough:

```
...supervisors(count()).sum()
```

What about something like this:

```
...supervisors(name,count(),...processes(count())).sum()
```

What would be the rules to use `sum()` in the spread relationship? These could be some options:
- Allow only one single selected aggregate (throw an error otherwise)
- Allow many aggregates and apply the `sum()` to each of them (throw error on non-aggregates)
- Apply to all the columns regardless and let PostgreSQL return the error
- Ignore non-aggregates and apply only on aggregated columns (does not throw an error, but PostgreSQL could return one if the aggregate type is a mismatch)

The last one could be a good approach, since it still allows spreading other columns... the implementation may not be trivial, though.

Reply:
To confirm we're in the same page, this is how I'm implementing the aggregated spreads. For instance, let's take this query:

```http
GET /factories?select=name,processes(process:name,count())
```

It returns something like:

```json
[
  {
    "name":"Factory A",
    "processes":[{"count": 1, "process": "Process A1"}, {"count": 1, "process": "Process A2"}]
  },
  {
    "name":"Factory B",
    "processes":[{"count": 1, "process": "Process B1"}, {"count": 1, "process": "Process B2"}]
  }
]
```

The aggregate, `count`  in this case, is done on a single row, that's why it always returns `1`. So, what would happen if we apply an spread operator to `processes`? There was a discussion that on to-one spreads, hoisting them to the top was the best solution (Q: https://github.com/PostgREST/postgrest/pull/2925#issuecomment-1720923400, A: https://github.com/PostgREST/postgrest/pull/2925#issuecomment-1720967590).  Now that we have to-many spreads, they'll be applied to the spread relationship itself, that is, count all of `processes` by factory (since I don't believe that a response like `count: [1,1] would be sensible in this case)`:

```http
GET /factories?select=name,...processes(process:name,count())
```

```json
[
  {
    "name":"Factory A",
    "processes":["Process A1", "Process A2"],
    "count": 2
  },
  {
    "name":"Factory B",
    "processes":["Process B1","Process B2"],
    "count": 2
  }
]
```

On nested to-many spreads this won't go to the most top one (like it does in to-one spreads), it will be also spread as arrays (like the one example I gave above https://github.com/PostgREST/postgrest/pull/3640#discussion_r1830230795):

```http
GET /factories?select=name,...processes(process:name,supervisors(supervisor:name,count()))
```

```json
[
  {
    "factory":"Factory B",
    "process":["Process B1", "Process B2"],
    "supervisor":[["Matt","Jeff"], ["Maria","Peter"]],
    "supervisor_count":[2,2]
  },
  "..."
]
```

Reply:
> I disagree. The [1, 1] is what I would expect here.

Yes, I agree. I didn't mean to publish the comment (just deleted it before reading your answer, sorry). I was still experimenting with the expected results and found out that it's expected to get the full `count()` **only when** not specifying an extra column, which makes sense.

Ignore my previous comment, I'll make a new example in a moment.